src/librustc/middle/check_match.rs

```diff
@@ -966,25 +966,23 @@ fn check_fn(cx: &mut MatchCheckCtxt,
     visit::walk_fn(cx, kind, decl, body, sp);
 
     for input in decl.inputs.iter() {
-        is_refutable(cx, &*input.pat, |pat| {
+        if let Some(pat) = refutable_pattern(cx, &*input.pat) {
             span_err!(cx.tcx.sess, input.pat.span, E0006,
                 "refutable pattern in function argument: `{}` not covered",
-                pat_to_string(pat)
+                pat_to_string(&*pat)
             );
-        });
+        }
         check_legality_of_move_bindings(cx, false, slice::ref_slice(&input.pat));
         check_legality_of_bindings_in_at_patterns(cx, &*input.pat);
     }
 }
 
-fn is_refutable<A, F>(cx: &MatchCheckCtxt, pat: &Pat, refutable: F) -> Option<A> where
-    F: FnOnce(&Pat) -> A,
-{
+fn refutable_pattern(cx: &MatchCheckCtxt, pat: &Pat) -> Option<P<Pat>> {
     let pats = Matrix(vec!(vec!(pat)));
     match is_useful(cx, &pats, &[DUMMY_WILD_PAT], ConstructWitness) {
         UsefulWithWitness(pats) => {
             assert_eq!(pats.len(), 1);
-            Some(refutable(&*pats[0]))
+            pats.into_iter().next()
```

Comment:
Could you keep this? The fact that we may create new `P<Pat>` shouldn't be exposed where possible.